ui: clusterviz: correct liveness status text on nodes

vilterp opened the pull request:
It used to always say "up for <duration>" no matter what the liveness status was. Now it will still say that if the node is live, and will say the status of the node otherwise.
Now:
![image](https://user-images.githubusercontent.com/7341/36818873-08024892-1cb5-11e8-9337-9929963a943c.png)

Would also be good to pipe in durations for other statuses (e.g. suspect), although this might be different enough to warrant a separate PR.

Fixes #23250 

Release note (admin ui change): clusterviz: correct liveness status text on nodes

Commit:
ui: clusterviz: correct liveness status text on nodes

It used to always say "up for <duration>" no matter what the liveness
status was. Now it will still say that if the node is live, and will say
the status of the node otherwise.

Release note (admin ui change): clusterviz: correct liveness status text on nodes

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23255)
<!-- Reviewable:end -->

Commit:
ui: clusterviz: show deadness duration on node component

This includes piping the liveness (which is different than the
LivenessStatus, confusingly) down all they way to the node view.

Also, proto imports got a little messy.

Release note (admin ui change): In the node map, show how long a node
has been dead.

vilterp commented:
Thoughts on whether it looks better upper case? I'm not sure.
![image](https://user-images.githubusercontent.com/7341/36819105-f0fe0d10-1cb5-11e8-89ab-5390a8afe89d.png)

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a nit

---

Reviewed 1 of 1 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/nodeCanvasContainer.tsx, line 34 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23255#-L6Y5yNBtHalOxgop_mt:-L6Y5yNBtHalOxgop_mu:br5652e) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f2aa69fa0edf57cce39b78c132bfdb28eb043a/pkg/ui/ccl/src/views/clusterviz/containers/map/nodeCanvasContainer.tsx#L34)):*
> ```XML
> import spinner from "assets/spinner.gif";
> 
> import { cockroach } from "src/js/protos";
> ```

Are these all ordered correctly?  Seems like they're not to me.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/nodeView.tsx, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23255#-L6Y5kyE5rYwq-BB1Lgo:-L6Y5kyE5rYwq-BB1Lgp:b-49tt14) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7b22631107094bfee794070908fd6b3f2a6beb/pkg/ui/ccl/src/views/clusterviz/containers/map/nodeView.tsx#L71)):*
> ```XML
>         return "unavailable";
>       default:
>         return ""; // idk man
> ```

Best code comment I've seen in a while.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23255#-:-L6Y6CRErscy6FYm_QZl:b-d5ftqp)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/nodeCanvasContainer.tsx, line 34 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23255#-L6Y5yNBtHalOxgop_mt:-L6YAurrtEDme3xVRkFh:b-l2g6qj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f2aa69fa0edf57cce39b78c132bfdb28eb043a/pkg/ui/ccl/src/views/clusterviz/containers/map/nodeCanvasContainer.tsx#L34)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Are these all ordered correctly?  Seems like they're not to me.
</blockquote></details>

Put it at the bottom to better match other files.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/nodeView.tsx, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23255#-L6Y5kyE5rYwq-BB1Lgo:-L6Y7enfkvGcTCTaj95Q:b-6jvxoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf7b22631107094bfee794070908fd6b3f2a6beb/pkg/ui/ccl/src/views/clusterviz/containers/map/nodeView.tsx#L71)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Best code comment I've seen in a while.
</blockquote></details>

Lol. I love exhaustivity checking of switches/pattern matches so much, and wish typescript had it (not too hard for enums, right?) Lets you get rid of this awkwardness.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23255)*
<!-- Sent from Reviewable.io -->

couchand commented:
@vilterp cherry pick please?

couchand commented on pkg/ui/ccl/src/views/clusterviz/containers/map/circleLayout.tsx:
```diff
@@ -14,13 +14,16 @@ import { getChildLocalities } from "src/util/localities";
 import { NodeView } from "./nodeView";
 import { LivenessStatus } from "src/redux/nodes";
+import { cockroach } from "src/js/protos";
+import Liveness = cockroach.storage.Liveness;
```
should this be `import`?  I'm surprised that even works?

vilterp replied:
Changed some of them to `type`, but for `NodeLivenessStatus`, it has to be `import`, otherwise the enum members are not available. (in this file: https://github.com/cockroachdb/cockroach/pull/23255/files#diff-85a6b40255d272377fc904208ce0f4ddR25)

couchand commented on pkg/ui/ccl/src/views/clusterviz/containers/map/circleLayout.tsx:
```diff
@@ -63,7 +66,8 @@ export class CircleLayout extends React.Component<CircleLayoutProps> {
                   node={node}
-                  liveness={this.props.liveness}
+                  livenessStatus={this.props.livenessStatus}
+                  liveness={this.props.liveness[node.desc.node_id]}
```
it's super confusing that these props are inconsistent...

vilterp replied:
Fixed them up to be consistent with the types — `livenessStatuses` is always `{ [id: string]: LivenessStatus }`; `livenesses` is `{ [id: string]: Liveness }`. Considered doing `livenessByNodeID` and `livnessStatusByNodeID`, but those are a little verbose.

I wish there was a single object that contained both, but that's not how the protobufs are organized, and I didn't feel like introducing a new type in TS…

couchand replied:
Yes, the Right Thing to Do (TM) is put them both on one thing, but probably not in the 2.0 timeframe.

couchand commented on pkg/ui/ccl/src/views/clusterviz/containers/map/nodeView.tsx:
```diff
@@ -45,23 +46,51 @@ export class NodeView extends React.Component<NodeViewProps> {
+      case NodeLivenessStatus.UNKNOWN:
+        return "unknown";
+      case NodeLivenessStatus.UNAVAILABLE:
+        return "unavailable";
```
probably better to call this "suspect" so it's consistent with elsewhere.

vilterp replied:
Just discovered https://github.com/cockroachdb/cockroach/blob/b47e107464d4a6ccf4de471e6ef76941fcf737cc/pkg/ui/src/redux/nodes.ts#L25-L34; should be using that here

couchand replied:
:+1:

couchand commented on pkg/ui/ccl/src/views/clusterviz/containers/map/nodeView.tsx:
```diff
@@ -45,23 +46,51 @@ export class NodeView extends React.Component<NodeViewProps> {
+      case NodeLivenessStatus.UNAVAILABLE:
+        return "unavailable";
+      default:
+        return ""; // idk man
```
probably better to call this "unknown"???

vilterp replied:
Also addressed by using `livenessNomenclature` function.

couchand reviewed (commented):
some followup thoughts, nothing critical except the user-visible one
